### PR TITLE
Add `GlobalLoopInvariantCodeMotionPass` to hoist constant `tensor.pack` from loops

### PR DESCRIPTION
**Pull request by pzread**
Add `GlobalLoopInvariantCodeMotionPass` (`--iree-global-opt-loop-invariant-code-motion`) to hoist loop invariants in global optimization stage.

As the first step we only limit the case on constant `tensor.pack %cst` in `scf.while` loop, which is a common case when enabling data-tiling. However the pass can be easily extended to support other ops and loops.

### CODE REVIEW

**Comment by pzread** [comment]
On some models this will trigger #16320 and need to patch a temporary fix #16363.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: `KD` refers to `Kernel Dispatch` so we may want to use LICM here?

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: needed here? all the methods below are either `static` or are within an anonymous namespace. We may want to move it to the very end, for the public method.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: the tree of producers is usually called backward slice, in case that name helps

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: "the nested ops" are hoistable? we don't have region op cases, right?

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Not sure I understand this check. It looks like this is checking if an operand is defined inside its operation itself?

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: I think we also have to check that the operation we want to hoist is immediately nested in the loop. I'm not sure we can hoist operations out of an nested if statement, for example.

**Review by dcaballe — COMMENTED** [review]
LGTM overall! I left a few initial comments. I think we may want to test multiple nested loop cases or just disable them for now. It's not clear to me we can hoist an operation that is nested in an if statement (zero-trip check).

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: When adding the hoistable ops at line 111, we only add immediately nested ops in the loop. I added a test to make sure we don't hoist ops from deeper than that.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Done

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Done

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Done

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Yes we don't have region ops for now, I wrote the code to also handle those cases.

I removed these logic and simplified the code to handle no region op for now

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Oh, I wrote this incorrectly.

It was inspired by the upstream LICM: https://github.com/llvm/llvm-project/blob/27726920e0366c99f5c940d304ea442515738974/mlir/lib/Transforms/Utils/LoopInvariantCodeMotionUtils.cpp#L40-L49 but it should check if the value is defined in the op we want to hoist.

For example:

```
scf.while {
  linalg.generic {
    %a = ...
    %b = arith.addf %a, %a
  }
}
```

`%b` is a loop invariant because `%a` is defined inside the `linalg.generic` we want to hoist.

I removed this check as we don't handle the ops with regions now

**Comment by pzread** [comment]
> LGTM overall! I left a few initial comments. I think we may want to test multiple nested loop cases or just disable them for now. It's not clear to me we can hoist an operation that is nested in an if statement (zero-trip check).

I added a new test to test loop-in-loop case (`@hoist_pack_op_with_zero_trip_check_in_outer_loop`). Currently we only hoist the op with no region so the loop invariants in the inner loop are hoisted but still stay inside the outer loop body.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: It would be good to add a comment about this there

**Review by dcaballe — APPROVED** [review]
Thanks for addressing the comments! LGTM! I'm approving but please make sure you get another approval from @benvanik or @MaheshRavishankar before landing.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Instead of recursing can we just use a work list here

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: This kind of lamda usage to recurse is very hard to follow. These all seem like they need to be worklists

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Reviewing, but left some comments to start

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: I am confused about the iteration order here. We are hoisting a backward slice but we are iterating the ops in a "forward manner" . I would expect to walk "backwards". Maybe use the post order, reverse iterator for walks to find the root of the op and the hoist the slice?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Added some comments to explain. The iteration order doesn't matter too much here because in this step we only search for hoistable root op and add them to `rootOpsToHoist` list. The next step below is to actually wrap the loop and hoist ops in `rootOpsToHoist` and their backward slice (I changed the actual hoisting step to run backward).

This way we can know there is at least one op in `rootOpsToHoist` before we actually wrap the loop.

Comment (dcaballe) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Not sure I follow... doesn't hoisting in a top-down order ensure that all the producers of an op (except `tensor.empty`) have been hoisted already and make the hoisting of the op simpler? I understood that hoisting a "backward slice" is only to hoist `tensor.empty`, which is an op we want to hoist only when its consumer is hoistable. I think hoisting backward slice shouldn't be doing complex hoistings of producer ops other than these "conditionally hoistable" ops, but I might be missing something.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: I was thinking about this but currently I still think it will be more difficult to implement with worklist.

The reason is that we want to know if the root op's operands can be hoisted together, so the same check also needs to be done on the operands, and so on, recursively. We only know the root op is hoistable after all check results returned from its operands, so the recursive works well here.

In my mind to get rid of recursive, we need a stack instead of worklist, means that the root op will push its operands into the stack and stay in the bottom until all its operands are checked. I can refactor with a stack but I feel like recursive is easier to understand.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: I refactor this to not do lambda recursion.

Regarding the worklist vs. recursive, IIUC the tricky part here is to hoist them in topological order. The recursive is following that naturally. To use a worklist we need to collect all ops and do a topological sorting first, which seems to be more complicated to me.

This code is also inspired by how IREE currently recursively hoists expression tree into globals: 

https://github.com/openxla/iree/blob/0c61f779bf9e56938a66625d214ab36fb85f684c/compiler/src/iree/compiler/Dialect/Util/Transforms/HoistIntoGlobals.cpp#L246-L257

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Added.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Had an offline discussion with @dcaballe . I changed the hoisting function `hoistOps` to use worklist to gather backward slices and sort all ops to be hoisted with `mlir::computeTopologicalSorting`, then hoist them in topological order.

This should resolve dependency issue and avoid recursion, while not making code too complicated. PTAL

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: I reworked the hoisting function into `hoistOps` and got rid of recursion (see https://github.com/openxla/iree/pull/16362#discussion_r1491749188). The iterating order is also done in topological order. Hopefully it is clearer now

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: typo: backward

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: -> SmallVectorImpl

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Cool!

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Done.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Done.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: I dont see why we are going back and forth. My expectation is that you do a forward analysis starting with the first operation of the loop and collect things that are hositable. If you go in program order, within the loop you have already traversed all the hoistable ops. If any of the operands are not defined by hoistable ops (or are defined outside the loop) then it is not hoistable. 
If you want to check whether `tensor.empty` must be hoisted (I dont follow the condition of why you want to hoist a tensor.empty only if its consumers are hoistable, but thats a separate question), you do a "hostiable analysis" above and then look to see if all consumers of `tensor.empty` are also marked hoistable. If not you drop it from the set. I am still not following why we need both forward and backward analysis.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Nice!

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Phew! much simpler! I think I just have one more question.

**Comment by MaheshRavishankar** [comment]
Had an offline discussion with @dcaballe . Is it possible to remove the "conditional" behavior. Basically it seems to me that we should always hoist `tensor.empty` outside the loop if we can (I cant think of a reason not to) especially at this level of the program. @pzread apparently there are some concerns w.r.t this you have?

**Comment by pzread** [comment]
> Had an offline discussion with @dcaballe . Is it possible to remove the "conditional" behavior. Basically it seems to me that we should always hoist `tensor.empty` outside the loop if we can (I cant think of a reason not to) especially at this level of the program. @pzread apparently there are some concerns w.r.t this you have?

My original thought was just to be more conservative, like not hoisting `tensor.empty` (which potentially becomes a buffer allocation) outside the loop while its user is still inside the loop. I assume we want to keep them closer. But I'm happy to make this simpler if those are not concerns.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: As the comment https://github.com/openxla/iree/pull/16362#issuecomment-1947529788, I dropped the difference between conditionally and unconditionally hoistable op for now.

The code is refactored to 1. Forward iteration to find hoistable op and build the `opsToHoist` list 2. Wrap the loop if there is any op to hoist 3. Hoist `opsToHoist` in order. This also gets rid of all recursions

**Comment by MaheshRavishankar** [comment]
> > Had an offline discussion with @dcaballe . Is it possible to remove the "conditional" behavior. Basically it seems to me that we should always hoist `tensor.empty` outside the loop if we can (I cant think of a reason not to) especially at this level of the program. @pzread apparently there are some concerns w.r.t this you have?
> 
> My original thought was just to be more conservative, like not hoisting `tensor.empty` (which potentially becomes a buffer allocation) outside the loop while its user is still inside the loop. I assume we want to keep them closer. But I'm happy to make this simpler if those are not concerns.

It's probably better to hoist the empty anyway. Otherwise you are basically creating a new tensor everytime. We might still run into issues, but in terms of program representation keeping it out makes more sense than inside the loop. So we can start with that

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: All tests passed and there is no regression. It is ready to be reviewed again

**Review by dcaballe — APPROVED** [review]
Thanks for bearing with us! LGTM

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: I think you actually need to walk the ops per region... Just iterating might not always work. In this case we have only one block, so it works. So either check for a single block and dont do anything if it isnt a single block region, then you can just iterate over ops in the block.
If you want to handle multiple blocks, you need to walk the IR to make sure you are getting producers before consumers.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: I am surprised this works. You need to use a `MapVector` to ensure that things are in insertion order. `SmallDenseMap` does not give you any gaurantees (and  can cause intermittent failures).

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: You can actually dont need a map here. You just need a `SetVector`. If it is in the set it is hoistable, if it is not then its not hoistable. `Map<Operation *, bool>` is effectively a `Set`.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp:84]: Better to leave a comment here that it this transformation shouldnt actually clone the operations. Basically the state of whats hostable that is stored using `Operation *` is still valid with this transformation.

This is still a footgun though. But would be better if you could hoist first (and get done with the state that `opsToHoist` is carrying before you do this transformation.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Thanks for making the logic for hoisting simpler. There are still a few issues with the mechanism of how the state is being carried. I left comments below.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: We can add checks but current SCF loops can only have one block per region, I think, and we don't support hoisting from ops that are nested into other region ops. The only loop we gather right now (`scf.while`) has two regions, which are listed in topological order. I don't think we can do much from the `LoopLikeInterfaceAPI` right now. We may want to revisit this as we add more loop types, if that ever happens.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: `opsToHoist` is a `SmallVector` but, yes, we can turn it into a `SetVector` and remove the other map we use.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: > We can add checks but current SCF loops can only have one block per region, I think, and we don't support hoisting from ops that are nested into other region ops. The only loop we gather right now (`scf.while`) has two regions, which are listed in topological order. I don't think we can do much from the `LoopLikeInterfaceAPI` right now. We may want to revisit this as we add more loop types, if that ever happens.

LoopLikeOpInterface doesn't check for a single block does it? This is written on the interface. If you don't explicitly abort on multiple blocks it can cause crashes for ops that implement the interface and have multiple blocks

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: > `opsToHoist` is a `SmallVector` but, yes, we can turn it into a `SetVector` and remove the other map we use.

Ah ok, I missed that part, but SetVector seems simpler.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Done.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: Changed to SetVector

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp]: For simplicity now, I added a check to make sure loop's region has only single block.

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (pzread) [reply compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp:84]: Added a comment and an assertion below to give us a hint when this assumption no longer holds

File: compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp:79]: Nit: I just only realized you need to do this only for while loops, cause you dont need to do this with an `scf.for`? Wasnt sure why we are doing this before hoisting and only now realized that this is because of `while` loop semantics.
This makes sense now.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks @pzread for bearing with me. This seems much simpler and hopefully still met your use case. Thanks for pushing on this!

**Comment by pzread** [comment]
Thanks for the review

**Pull request merged**
